### DISCUSSION TITLE
Undefined reference to `Ipopt::IpoptApplication::IpoptApplication(bool, bool)'

### DISCUSSION THREAD

**Discussion by Jackhe96** [coin-or/CppAD#104] [General]
I am completely new to cppad and I apologize if this is a stupid question. I have been trying to install cppad and ipopt for a day now and it's getting really frustrating as I don't know how to solve this problem. I am trying to use Cppad in my program but I run into the error:

In function `void CppAD::ipopt::solve<CppAD::vector<double>, (anonymous namespace)::FG_eval>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, CppAD::vector<double> const&,CppAD::vector<double> const&, CppAD::vector<double> const&, CppAD::vector<double> const&, CppAD::vector<double> const&, (anonymous namespace)::FG_eval&, CppAD::ipopt::solve_result<CppAD::vector<double> >&)':
ipopt_test.cpp:(.text+0xe0a): undefined reference to `Ipopt::IpoptApplication::IpoptApplication(bool, bool)'
collect2: error: ld returned 1 exit status

I can't find many resources online of people having the same problem. To check if it is my program I tried running the following [example]( https://www.coin-or.org/CppAD/Doc/ipopt_nlp_get_started.cpp.htm) but I get the same error. Wondering if you guys know about it

**Comment by bradbell** [comment]
It appears that you are not including the ipopt library on your in the
command that is creating the executable.

It might help to run a simple ipopt example, that does not is CppAD, to
ensure it is properly installed and the linking works.

**Comment by bradbell** [comment]
Please close this issue if the answer above solved this problem.

**Comment by lxjlu** [comment]
I also have the same question. How do you solve it in the end? Thanks!

**Comment by Jackhe96** [comment]
@lxjlu It was my fault, I was using ROS and the cmakelists.txt file was not configured correctly

**Reply by HHendy69** [reply]
So what was the right configuration in CMakeList looks like?

**Comment by lxjlu** [comment]
> @lxjlu It was my fault, I was using ROS and the cmakelists.txt file was not configured correctly

I also used it for the first time, and I understand whether it is a problem in the link section. I am directly g++ -o example example.cpp. Is there any problem, thank you.

**Comment by bradbell** [comment]
I assume you are using Ipopt and have installed it.

If on your cmake command line you choose cmake_verbose_makefile eqaul to YES
https://coin-or.github.io/CppAD/doc/cmake.htm#cmake_verbose_makefile

and you also specify ipopt_prefix
https://coin-or.github.io/CppAD/doc/cmake.htm#package_prefix

then during the make check
https://coin-or.github.io/CppAD/doc/cmake.htm#make%20check

you should see how CppAD links in ipopt.

**Comment by lxjlu** [comment]
> I assume you are using Ipopt and have installed it.
> 
> If on your cmake command line you choose cmake_verbose_makefile eqaul to YES
> https://coin-or.github.io/CppAD/doc/cmake.htm#cmake_verbose_makefile
> 
> and you also specify ipopt_prefix
> https://coin-or.github.io/CppAD/doc/cmake.htm#package_prefix
> 
> then during the make check
> https://coin-or.github.io/CppAD/doc/cmake.htm#make%20check
> 
> you should see how CppAD links in ipopt.

Thanks, I will check it carefully.

**Comment by bradbell** [comment]
It will be easier to see if you only check to ipopt_solve examples; i.e., use
make check_example_ipopt_solve

**Comment by tu-trinh** [comment]
@Jackhe96 Hi, I'm running into the exact same error. How did you fix your cmakelists.txt?
I just freshly installed ipopt and #included cppad/ipopt/solve.hpp as well as Interfaces/IpIpoptApplication.hpp and using Ipopt::IpoptApplication, which I believe is where IpoptApplication comes from.

**Comment by bradbell** [comment]
@tu-trinh  By the exact same error do you mean ?
```
undefined reference to Ipopt::IpoptApplication::IpoptApplication(bool, bool)'
collect2: error: ld returned 1 exit status
```

**Comment by DanielrdMorais** [comment]
I have the same problem:
undefined reference to `Ipopt::IpoptApplication::IpoptApplication(bool, bool)'
collect2: error: ld returned 1 exit status

**Comment by bradbell** [comment]
Is the undefined reference during the testing of CppAD ?
Perhaps it would help to look at the following discussion: https://github.com/coin-or/CppAD/discussions/103

**Reply by DanielrdMorais** [reply]
This is the full error:

 in function `void CppAD::ipopt::solve<CppAD::vector<double>, FG_eval>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, CppAD::vector<double> const&, CppAD::vector<double> const&, CppAD::vector<double> const&, CppAD::vector<double> const&, CppAD::vector<double> const&, FG_eval&, CppAD::ipopt::solve_result<CppAD::vector<double> >&)':
MPC.cpp:(.text._ZN5CppAD5ipopt5solveINS_6vectorIdEE7FG_evalEEvRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKT_SF_SF_SF_SF_RT0_RNS0_12solve_resultISD_EE[_ZN5CppAD5ipopt5solveINS_6vectorIdEE7FG_evalEEvRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKT_SF_SF_SF_SF_RT0_RNS0_12solve_resultISD_EE]+0x182): undefined reference to `Ipopt::IpoptApplication::IpoptApplication(bool, bool)'
collect2: error: ld returned 1 exit status

Thanks in advance

**Reply by DanielrdMorais** [reply]
Solved, this was missing in my CMakeLists.txt:

target_link_libraries(main ipopt)

**Reply by chenfei8888** [reply]
> L

hello，target_link_libraries(main ipopt)    is used in my cmake file，but it still hasn't been solved.


my cmakelist is:

cmake_minimum_required(VERSION 3.0.0)
project(ipopt_tests VERSION 0.1.0)
list(APPEND CMAKE_MODULE_PATH "/usr/local/lib/cmake")
add_definitions(-std=c++11 -O3)

set(CXX_FLAGS "-Wall")
set(CMAKE_CXX_FLAGS, "${CXX_FLAGS}")

set(sources src/ipopt_tests.cpp)

include_directories(/usr/include)
link_directories(/usr/local/lib /usr/local/lib/packconfig)
include_directories(/usr/include/eigen3)

add_executable(ipopt_test ${sources})

target_link_libraries(ipopt_test ipopt)

**Reply by chenfei8888** [reply]
The problem has been solved, the vscode environment is not working.
launch.js should be set up

**Reply by bradbell** [reply]
Can accomplish the necessary cmake setting using the distributed version of CMakeLists.txt and the instructions on ?
https://coin-or.github.io/CppAD/doc/cmake.htm

**Reply by chenfei8888** [reply]
Yes, I've set up CMakeList.txt. Thank you very much.

**Reply by chenfei8888** [reply]
My question has been solved by CMakeList and .launch file

**Comment by HHendy69** [comment]
I am still getting the following error 
`In function `void CppAD::put_check_for_nan<double>(CppAD::vector<double> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)':
MPC.cpp:(.text._ZN5CppAD17put_check_for_nanIdEEvRKNS_6vectorIT_EERNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE[_ZN5CppAD17put_check_for_nanIdEEvRKNS_6vectorIT_EERNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE]+0x64): undefined reference to `CppAD::local::temp_file[abi:cxx11]()'`

I ain't checking out where is the problem
this how it looks like in my CMake
project(ipopt_tests)

cmake_minimum_required (VERSION 3.5)

add_definitions(-std=c++11)

set(CXX_FLAGS "-Wall")
set(CMAKE_CXX_FLAGS "${CXX_FLAGS}")

set(sources src/ipopt_test.cpp src/main.cpp)

include_directories(src/Eigen-3.3)
include_directories(/usr/local/include)
link_directories(/usr/local/lib)

add_executable(ipopt_tests ${sources})

target_link_libraries(ipopt_tests ipopt python2.7)

**Reply by rtheta02043** [reply]
Same problem, is this solved @hichamhendy ?

**Reply by HHendy69** [reply]
Yeah. The problem lied in the mathematical formulation of the optimisation problem. I corrected it and got it running.

**Reply by rtheta02043** [reply]
Okay... thanks!

**Reply by rtheta02043** [reply]
@hichamhendy I am guessing you are referring to MPC.cpp file. Which part of the MPC formulation is causing this problem could you explain please?

**Reply by HHendy69** [reply]
no I am using that for quadcopter

**Comment by bradbell** [comment]
I just fixed a problem that came up when building the ipopt tests and ipopt is the only optional package. (Testing this was motivated by your message above); see 03-04 on
https://cppad.readthedocs.io/en/latest/whats_new_23.html#whats-new-23-mm-dd-03-04

Please test the current master and see if it fixes your problem.

**Comment by zhumingpassional** [comment]
I also encountered the same problem.

@hichamhendy could you pls share the simple demo codes with me?

**Comment by bradbell** [comment]
Try the following on your machine and see if you get an error message ? 

1. git clone https://github.com/coin-or/CppAD.git cppad.git
2. cd cppad.git
3. bin/get_ipopt.sh
4. export PKG_CONFIG_PATH="$(pwd)/build/prefix/lib/pkgconfig"
5. cd build
6. cmake -D include_ipopt=true ..
7. make check_example_ipopt_solve

I did not get any errors and got the final result:

    Begin test group example/ipopt_solve
    get_started         OK
    ode_inverse         OK
    retape              OK
    memory_leak         OK
    All 4 tests passed.
    End test group example/ipopt_solve
    [100%] Built target check_example_ipopt_solve
